### DISCUSSION TITLE
Is there a way for CMake (or Visual Studio) to ignore the system-wide vcpkg integration?

### DISCUSSION THREAD

**Discussion by guenterberner** [microsoft/vcpkg#30252] [Q&A] [answered]
I have C++ "prototyping" projects on my machine that I usually create from Visual Studio using "New Project...".

For these projects it is very handy to have the "vcpkg Integrate Install" command. Any third party library can be used in any project out of the box with no overhead.

On the other hand, I have some projects that are scripted with CMake and that pull their third-party libraries from a Maven repository - and shouldn't use anything from the local vcpkg installation.

Every time CMake generates the VS project files, they automatically use the vcpkg installation and this ultimately results in linker chaos because it partly uses the vcpkg libs and also the maven libs (which it cannot find in vcpkg).

That's why I have to open the project/solution files with VS _each time_ after creating them and switch off vcpkg manually.

So is there a way to tell CMake that Visual Studio should "ignore" the system-wide vcpkg integration?

Thank you.

**Comment by Osyotr** [comment]
Not a direct answer to your question, but in VS2019+ you can skip generation of VS project files and open folder with CMakeLists.txt directly.

**Reply by guenterbe** [reply]
Unfortunately, this is not an option for the non-vcpkg based projects. CMake generation of build files and the build itself are controlled from Maven. This means that I have to interrupt the build, launch VS and edit all projects within the solution to not use vcpkg and then restart the build.
Nevertheless many thanks.

**Reply by Osyotr** [reply]
Maybe also use Ninja generator instead of Visual Studio generator?

**Reply by guenterbe** [reply]
That also does not work. We need the VS project files because we use VS as a working environment.

**Reply by Osyotr** [reply]
Hmm if you can control how `msbuild` is invoked, you can try passing `/p:VcpkgEnabled=false`
(or set VS_GLOBAL_VcpkgEnabled "false"` target property) https://cmake.org/cmake/help/latest/prop_tgt/VS_GLOBAL_variable.html

**Reply by guenterbe** [reply]
I give everything but I don't want to succeed. As soon as vcpkg is available globally (via "vcpkg integrate install"), VS shows me "Use Vcpkg Yes" in the project configuration.
I have no idea what I'm doing wrong.

So far I have tried:
set_property(GLOBAL PROPERTY VS_GLOBAL_VcpkgEnabled false)
set(VS_GLOBAL_VcpkgEnabled false)
set_target_properties(${TARGET} PROPERTIES VS_GLOBAL_VcpkgEnabled false)

**Reply by netcorefan1** [reply]
I faced the same issue and the closest I could get it is by creating a `Directory.build.props` at the root level:
```
<Project>
    <PropertyGroup>
        <VcpkgEnabled>false</VcpkgEnabled>
    </PropertyGroup>
</Project>
```
It works, but until a certain degree. If projects are automatically generated and contains other `Directory.build.props`, then `VcpkgEnabled` get overridden and vcpkg becomes enabled for the related project. I still have to find out how to get around of this.
Passing `/p:VcpkgEnabled=false` to `msbuild` has the same issue: there is no guarantee that the property will be propagated and some projects could still end up with vcpkg enabled.
As of now the only thing I could think of to solve the issue is patching the source code. There must be an easier, reliable and less intrusive way!

**Reply by guenterbe** [reply]
If vcpkg user wide integration is turned on via command line, _manually_ turning off Vcpkg support in VS project/solution configuration results in an .vcxproj entry like
```
  <PropertyGroup Label="Vcpkg">
    <VcpkgEnabled>false</VcpkgEnabled>
  </PropertyGroup>
```

This seems to work (at least on my machine). 

I just can't get CMake to _automatically_ create this entry when the VS/msbuild project is generated. I've tried all the known and suggested CMake entries above, but none of them creates such an entry in the .vcxproj file.

Btw, I have tried CMake version 3.25.1 and also 3.25.1-msvc1

**Reply by Neumann-A** [reply]
> I just can't get CMake to automatically create this entry when the VS/msbuild project is generated. 

normally vcpkg should already deactivate the integration via:

https://github.com/microsoft/vcpkg/blob/1271068e139c1fc30bae405c0bca0e379e155bd2/scripts/buildsystems/vcpkg.cmake#L650-L651

**Reply by guenterbe** [reply]
vcpkg integration state is enabled via

`vcpkg integrate install`

This is my stripped down CMakeLists.txt

```
cmake_minimum_required(VERSION 3.25)
set_property(GLOBAL PROPERTY VS_GLOBAL_VcpkgEnabled false)
set(VS_GLOBAL_VcpkgEnabled false)
set_target_properties(${TARGET} PROPERTIES VS_GLOBAL_VcpkgEnabled false)
project(no_vcpkg)
add_executable(no_vcpkg src/main.cpp)
```

and I always end up with vcpkg enabled in the generated vcxproj. I am running out of ideas.

**Reply by netcorefan1** [reply]
Yes, `vcpkg integrate install` is how integration is installed. It should be intended just like an install, not something that can be enabled and disabled on demand (otherwise they would have called it `vcpkg integrate enable/disable`). Avoid hackish solutions like `vcpkg integrate remove && MsbuidlThis & vcpkg integrate install`.
I didn't had much luck by changing CMakeLists.txt and touching the build system is something that I'm trying to avoid. If you're building from a cloned repo you can never be sure that it will break your code when there are changes to the source code.



> If vcpkg user wide integration is turned on via command line, _manually_ turning off Vcpkg support in VS project/solution configuration results in an .vcxproj entry like
> 
> ```
>   <PropertyGroup Label="Vcpkg">
>     <VcpkgEnabled>false</VcpkgEnabled>
>   </PropertyGroup>
> ```
> 
> This seems to work (at least on my machine).
> 
> I just can't get CMake to _automatically_ create this entry when the VS/msbuild project is generated. I've tried all the known and suggested CMake entries above, but none of them creates such an entry in the .vcxproj file.
> 
> Btw, I have tried CMake version 3.25.1 and also 3.25.1-msvc1

Exactly. If you set this property to false in a `Directory.build.props` file placed at the root of your solution, all the generated projects should automatically pickup your property and you should see `vcpkg enabled: false` in each project. You don't need to touch anything in the generated projects. It should work as long as the generator don't create a `Directory.build.props` in one of the generated project. In such case the project will take properties from the nearest `Directory.build.props` and since it doesn't contain any VcpkgEnabled property, vcpkg will be enabled for that project (the lack of this property is interpreted as `vcpkg enabled: true`). There must be a way to enforce this property propagation in all subfolders, but I still have to find out how.



> > I just can't get CMake to automatically create this entry when the VS/msbuild project is generated.
> 
> normally vcpkg should already deactivate the integration via:
> 
> https://github.com/microsoft/vcpkg/blob/1271068e139c1fc30bae405c0bca0e379e155bd2/scripts/buildsystems/vcpkg.cmake#L650-L651

I may be wrong, but this can only work for the starting target. In the case of generated projects, each one can have its own CMakeList. The same action should be performed on each CMakeList which is not possible in case of autogenerated projects.

Normally, the generation of the projects and the compilation of that projects are two separate commands. One could do something like this which should requires no more than a few lines of powershell code:
```
cmake configure
foreach (*.vcxproj in GeneratedProjectsFolder) {
  AddXmlProperty(VcpkgEnabled, false)
}
cmake build
```
This seems to me the cleanest and error prone solution. It allows more control because one can simply disable vcpkg in all the projects contained in a target folder or keep it enabled for the ones that match a specific criteria.

**UPDATE**: I've just realized that this will only work when building outside the ide, but if we hook the `pre-build` event and run the task there, it should work on both sides. However I'm not sure it is safe to change projects files while a build process is already started. May be is better an `add_custom_command` on CMake post build generation event.

**Reply by netcorefan1** [reply]
@guenterberner Did you tried to see if it worked my idea above? Now I'm involved in fixing other troubles with my project. Once done I will also have to deal with this problem because vcpkg automatic linkage cause build failure. If you'll be still interested I will post the code.

**Reply by guenterbe** [reply]
@netcorefan1 First of all, thank you very much for your extensive help.

1.
Unfortunately, the Powershell-based approach is out of the question. (Our build runs on Windows + Linux, is driven by Maven on both platforms (which in turn calls CMake to generate the C++ build files, download and install the libraries from a Maven repository and use msbuild to build the projects, etc.). So I would have to add custom steps in the maven pom files to manipulate the vcxproj files, which is even more pita than expanding a CMakeLists.txt)

2.
I've tried the Directory.build.props approach in a test project - but got an error from cmake when generatign the build files

```
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.19045.
CMake Error at CMakeLists.txt:3 (project):
  Failed to run MSBuild command:

    C:/Program Files/Microsoft Visual Studio/2022/Professional/MSBuild/Current/Bin/amd64/MSBuild.exe

  to get the value of VCTargetsPath:

    MSBuild version 17.5.1+f6fdcf537 for .NET Framework
    Build started 05.04.2023 12:05:11.
    Included response file: C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\MSBuild.rsp

    Project "W:\dev\vcpkg-test\CMakeFiles\3.23.0-rc2\VCTargetsPath.vcxproj" on node 1 (default targets).
    W:\dev\vcpkg-test\Directory.Build.props(1,1): error MSB4068: The element <PropertyGroup> is unrecognized, or not supported in this context. [W:\dev\vcpkg-test\CMakeFiles\3.23.0-rc2\VCTargetsPath.vcxproj]
    Done Building Project "W:\dev\vcpkg-test\CMakeFiles\3.23.0-rc2\VCTargetsPath.vcxproj" (default targets) -- FAILED.
```

**Reply by netcorefan1** [reply, marked as answer]
> Unfortunately, the Powershell-based approach is out of the question. (Our build runs on Windows + Linux, is driven by Maven on both platforms (which in turn calls CMake to generate the C++ build files, download and install the libraries from a Maven repository and use msbuild to build the projects, etc.). So I would have to add custom steps in the maven pom files to manipulate the vcxproj files, which is even more pita than expanding a CMakeLists.txt)

Mhhh... I have no experience on Maven, so I can't help here. Did you tried the `add_custom_command` approach?

> I've tried the Directory.build.props approach in a test project - but got an error from cmake when generatign the build files
> 
> ```
> -- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.19045.
> CMake Error at CMakeLists.txt:3 (project):
>   Failed to run MSBuild command:
> 
>     C:/Program Files/Microsoft Visual Studio/2022/Professional/MSBuild/Current/Bin/amd64/MSBuild.exe
> 
>   to get the value of VCTargetsPath:
> 
>     MSBuild version 17.5.1+f6fdcf537 for .NET Framework
>     Build started 05.04.2023 12:05:11.
>     Included response file: C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64\MSBuild.rsp
> 
>     Project "W:\dev\vcpkg-test\CMakeFiles\3.23.0-rc2\VCTargetsPath.vcxproj" on node 1 (default targets).
>     W:\dev\vcpkg-test\Directory.Build.props(1,1): error MSB4068: The element <PropertyGroup> is unrecognized, or not supported in this context. [W:\dev\vcpkg-test\CMakeFiles\3.23.0-rc2\VCTargetsPath.vcxproj]
>     Done Building Project "W:\dev\vcpkg-test\CMakeFiles\3.23.0-rc2\VCTargetsPath.vcxproj" (default targets) -- FAILED.
> ```

To me sounds like a file containing invalid characters. Try with a very basic file like the below (I don't remember if mine was with label, so try with and without and see what works):
```
<Project>
    <PropertyGroup Label="Vcpkg">
        <VcpkgEnabled>false</VcpkgEnabled>
    </PropertyGroup>
</Project>
```
If you're using non-SDK projects you may need some [changes](https://github.com/dotnet/samples/blob/main/Directory.Build.props). You could also try with `Directory.build.targets` (just rename the extension, no other change).

**Reply by guenterberner** [reply]
That's how it works!

File: Directory.build.props
Location: next to CMakeLists.txt

Content:
```
<Project>
    <PropertyGroup Label="Vcpkg">
        <VcpkgEnabled>false</VcpkgEnabled>
    </PropertyGroup>
</Project>
```

Thank you very much.

**Reply by netcorefan1** [reply]
I'm very glad to have helped you. Yes, it should be at the root where there is the CMakeLists.txt

> It should work as long as the generator don't create a `Directory.build.props` in one of the generated projects folder. In such case the project will take properties from the nearest `Directory.build.props` and since it doesn't contain any VcpkgEnabled property, vcpkg will be enabled for that project.

Remember to check all the folders to ensure that there aren't other auto generated `Directory.build.props` !
